zend_abstract_interface/config/config_ini.c

```diff
@@ -0,0 +1,140 @@
+#include "config_ini.h"
+
+#include <assert.h>
+#include <main/php.h>
+#include <stdbool.h>
+#include <string.h>
+
+#include "config.h"
+
+#define UNUSED(x) (void)(x)
+
+static void (*env_to_ini_name)(zai_string_view env_name, zai_config_ini_name *ini_name);
+
+static zend_array config_ini_name_map = {0};
+
+static bool zai_config_generate_ini_name(zai_string_view name, zai_config_ini_name *ini_name) {
+    ini_name->len = 0;
+    *ini_name->ptr = '\0';
+
+    env_to_ini_name(name, ini_name);
+
+    return ini_name->len && *ini_name->ptr;
+}
+
+bool zai_config_get_ini_value(zai_string_view name, zai_env_buffer buf) {
+    if (!env_to_ini_name) return false;
+
+    zai_config_ini_name ini_name;
+    if (!zai_config_generate_ini_name(name, &ini_name)) return false;
+
+    zend_bool exists;
+    char *raw_value = zend_ini_string_ex(ini_name.ptr, ini_name.len, /* orig */ 0, &exists);
+    if (!exists || !raw_value || !*raw_value) return false;
+
+    size_t len = strlen(raw_value);
+    // TODO Log this message
+    assert((len < buf.len) && "INI value is greater than the buffer size");
+    if (len >= buf.len) return false;
+
+    strncpy(buf.ptr, raw_value, len);
+    buf.len = len;
+
+    return true;
+}
+
+static bool zai_config_get_id_by_ini_name(zai_string_view name, zai_config_id *id) {
+    assert(name.ptr && name.len && id);
+    if (!GC_REFCOUNT(&config_ini_name_map)) {
+        assert(false && "INI name map not initialized");
+        return false;
+    }
+
+    zval *zid = zend_hash_str_find(&config_ini_name_map, name.ptr, name.len);
+    if (zid && Z_TYPE_P(zid) == IS_LONG) {
+        *id = Z_LVAL_P(zid);
+        return true;
+    }
+
+    return false;
+}
+
+static ZEND_INI_MH(ZaiConfigOnUpdateIni) {
+    UNUSED(mh_arg1);
+    UNUSED(mh_arg2);
+    UNUSED(mh_arg3);
+
+    /* Ignore calls that happen before runtime (e.g. the default INI values on MINIT). System values are obtained on
+     * first-time RINIT. */
+    if (stage != PHP_INI_STAGE_RUNTIME) return SUCCESS;
+
+    zai_config_id id;
+    zai_string_view name = {.len = ZSTR_LEN(entry->name), .ptr = ZSTR_VAL(entry->name)};
+
+    if (!zai_config_get_id_by_ini_name(name, &id)) {
+        // TODO Log cannot find ID
+        return FAILURE;
+    }
+
+    zval tmp;
+    ZVAL_UNDEF(&tmp);
+    zai_string_view value_view = {.len = ZSTR_LEN(new_value), .ptr = ZSTR_VAL(new_value)};
+
+    if (zai_config_decode_value(value_view, memoized_entires[id].type, &tmp, /* persistent */ false)) {
+        zai_config_update_runtime_config(id, &tmp);
```

Comment:
Do we assume that every config is safe to be changed at any time? … or maybe it's safe but has no effect.
I think in these cases we should forbid changing.

Reply:
As an example: `DD_TRACE_RETAIN_THREAD_CAPABILITIES` should really be `PHP_INI_SYSTEM` as it's a startup-only config, it just does not make sense to toggle that at runtime.

Reply:
> or maybe it's safe but has no effect.

Correct. As we briefly discussed a bit earlier, some of the cases where the config value was set too late could possibly have workarounds; (e.g. setting `DD_TRACE_GENERATE_ROOT_SPAN=0` at runtime could drop the existing spans.)

I guess that might mean we should consider a callback or something so that the tracer could subscribe to config changes?

> I think in these cases we should forbid changing.

How would ZAI config know it was too late for a particular value to be set? Perhaps if we add a callback to pass the decoded value to the tracer before the runtime config is updated, ZAI config can check the return value of the callback to determine if the value should be updated?

Reply:
We have to tell ZAI config. Trivially the "too late" part is usually PHP_INI_SYSTEM vs PHP_INI_ALL. For sure, we could be more granular here, but I don't think this is worth it.

Reply:
If we want more flexibility, especially for cases like root span, yes, then we need a callback.